File: src/lib/repositories/vlogsRepository.ts
import { run, getAll } from '@/lib/db';
import type { SavedVlog } from '@/types';
import * as FileSystem from 'expo-file-system/legacy';
import { CONFIG } from '@/config';

export interface VlogRow {
    id: string;
    file_path: string;
    date_str: string;
    timestamp: number;
    duration_sec: number;
    file_size_bytes: number;
    thumbnail_path: string | null;
    compression_preset: string | null;
    original_file_size_bytes: number | null;
    compression_pending: number;
}

function rowToVlog(row: VlogRow): SavedVlog {
    const videoFileName = row.file_path.substring(row.file_path.lastIndexOf('/') + 1);
    const correctedFilePath = `${FileSystem.documentDirectory}${CONFIG.VLOG_STORAGE_DIR}${videoFileName}`;

    let correctedThumbnailPath = row.thumbnail_path ?? undefined;
    if (correctedThumbnailPath) {
        const thumbFileName = correctedThumbnailPath.substring(correctedThumbnailPath.lastIndexOf('/') + 1);
        correctedThumbnailPath = `${FileSystem.documentDirectory}vlog_thumbnails/${thumbFileName}`;
    }

    return {
        id: row.id,
        filePath: correctedFilePath,
        dateStr: row.date_str,
        timestamp: row.timestamp,
        durationSec: row.duration_sec,
        fileSizeBytes: row.file_size_bytes,
        thumbnailPath: correctedThumbnailPath,
        compressionPreset: row.compression_preset ?? undefined,
        originalFileSizeBytes: row.original_file_size_bytes ?? undefined,
        compressionPending: !!row.compression_pending,
    };
}

export async function getAllVlogs(): Promise<SavedVlog[]> {
    const rows = await getAll<VlogRow>(`SELECT * FROM vlogs ORDER BY timestamp DESC`);
    return rows.map(rowToVlog);
}

export async function insertVlog(vlog: SavedVlog): Promise<void> {
    await run(
        `INSERT INTO vlogs (id, file_path, date_str, timestamp, duration_sec, file_size_bytes, thumbnail_path, compression_preset, original_file_size_bytes, compression_pending)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
        [
            vlog.id,
            vlog.filePath,
            vlog.dateStr,
            vlog.timestamp,
            vlog.durationSec,
            vlog.fileSizeBytes,
            vlog.thumbnailPath ?? null,
            vlog.compressionPreset ?? null,
            vlog.originalFileSizeBytes ?? null,
            vlog.compressionPending ? 1 : 0,
        ],
    );
}

export async function updateVlog(id: string, updates: Partial<SavedVlog>): Promise<void> {
    const fields: string[] = [];
    const values: (string | number | null)[] = [];

    if (updates.filePath !== undefined) {
        fields.push('file_path = ?');
        values.push(updates.filePath);
    }
    if (updates.dateStr !== undefined) {
        fields.push('date_str = ?');
        values.push(updates.dateStr);
    }
    if (updates.timestamp !== undefined) {
        fields.push('timestamp = ?');
        values.push(updates.timestamp);
    }
    if (updates.durationSec !== undefined) {
        fields.push('duration_sec = ?');
        values.push(updates.durationSec);
    }
    if (updates.fileSizeBytes !== undefined) {
        fields.push('file_size_bytes = ?');
        values.push(updates.fileSizeBytes);
    }
    if (updates.thumbnailPath !== undefined) {
        fields.push('thumbnail_path = ?');
        values.push(updates.thumbnailPath ?? null);
    }
    if (updates.compressionPreset !== undefined) {
        fields.push('compression_preset = ?');
        values.push(updates.compressionPreset ?? null);
    }
    if (updates.originalFileSizeBytes !== undefined) {
        fields.push('original_file_size_bytes = ?');
        values.push(updates.originalFileSizeBytes ?? null);
    }
    if (updates.compressionPending !== undefined) {
        fields.push('compression_pending = ?');
        values.push(updates.compressionPending ? 1 : 0);
    }

    if (fields.length === 0) return;
    values.push(id);
    await run(`UPDATE vlogs SET ${fields.join(', ')} WHERE id = ?`, values);
}

export async function deleteVlog(id: string): Promise<void> {
    await run(`DELETE FROM vlogs WHERE id = ?`, [id]);
}

export async function deleteAllVlogs(): Promise<void> {
    await run(`DELETE FROM vlogs`);
}

File: mda_flutter/lib/data/database/repositories/vlogs_repository.dart
/// Vlogs repository — port of `src/lib/repositories/vlogsRepository.ts`.
/// Rows are path-rebased to the current device's documents directory so stale
/// absolute paths (after reinstall) heal automatically.
library;

import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';

import '../../models/saved_vlog.dart';
import '../db.dart';

class VlogsRepository {
  VlogsRepository({Future<String> Function()? documentsDirProvider})
      : _documentsDirProvider = documentsDirProvider ?? _defaultDocumentsDir;

  final Future<String> Function() _documentsDirProvider;

  static Future<String> _defaultDocumentsDir() async {
    final docs = await getApplicationDocumentsDirectory();
    return docs.path;
  }

  /// All vlogs, newest first — with paths rebased to the current sandbox.
  Future<List<SavedVlog>> getAllVlogs() async {
    final docs = await _documentsDirProvider();
    final rows = await getAll('SELECT * FROM vlogs ORDER BY timestamp DESC');
    return rows.map((row) => _rebased(row, docs)).whereType<SavedVlog>().toList();
  }

  /// Rebases file_path/thumbnail_path to `$docs/vlogs/` / `$docs/vlog_thumbnails/`.
  SavedVlog? _rebased(Map<String, Object?> row, String docs) {
    final vlog = SavedVlog.fromRow(row);
    if (vlog == null) return null;
    return SavedVlog(
      id: vlog.id,
      filePath: p.join(docs, 'vlogs', p.basename(vlog.filePath)),
      dateStr: vlog.dateStr,
      timestamp: vlog.timestamp,
      durationSec: vlog.durationSec,
      fileSizeBytes: vlog.fileSizeBytes,
      thumbnailPath: vlog.thumbnailPath != null
          ? p.join(docs, 'vlog_thumbnails', p.basename(vlog.thumbnailPath!))
          : null,
      compressionPreset: vlog.compressionPreset,
      originalFileSizeBytes: vlog.originalFileSizeBytes,
      compressionPending: vlog.compressionPending,
    );
  }

  Future<void> insertVlog(SavedVlog vlog) async {
    final row = vlog.toRow();
    final columns = row.keys.join(', ');
    final placeholders = List.filled(row.length, '?').join(', ');
    await run('INSERT INTO vlogs ($columns) VALUES ($placeholders)', row.values.toList());
  }

  static const _allowedColumns = {
    'file_path', 'file_size_bytes', 'thumbnail_path', 'compression_preset',
    'original_file_size_bytes', 'compression_pending',
  };

  Future<void> updateVlog(String id, Map<String, Object?> updates) async {
    final sets = <String>[];
    final values = <Object?>[];
    for (final entry in updates.entries) {
      if (!_allowedColumns.contains(entry.key)) continue;
      sets.add('${entry.key} = ?');
      values.add(entry.value);
    }
    if (sets.isEmpty) return;
    values.add(id);
    await run('UPDATE vlogs SET ${sets.join(', ')} WHERE id = ?', values);
  }

  Future<void> deleteVlog(String id) async {
    await run('DELETE FROM vlogs WHERE id = ?', [id]);
  }

  Future<void> deleteAllVlogs() async {
    await run('DELETE FROM vlogs');
  }
}
